File: FormacaoPhp/PHPAvancOOP/src/model/ContaCorrente.php
<?php

namespace src\model;

use src\exceptions\OperacaoNaoRealizadaException;
use src\logics\Validacao;
use src\exceptions\SaldoInsuficienteException;

class ContaCorrente
{
    private $titular;
    private $agencia;
    private $numero;
    private $saldo;

    public $totalSaquesNaoPermitidos;

    public static $operacoesNaoRealizadas;

    public static $taxaOperação;

    public static $totalDeContas;


    /* ----------- Getters and Setters ----------- */
    public function __get($attribute)
    {
        return $this->$attribute;
    }

    public function __set($attribute, $valor)
    {
        if ($attribute == 'saldo' || $attribute == 'titular')
            return;
        $this->$attribute = $valor;
    }
    /* ----------- Getters and Setters -----------*/

    public function __construct($titular, $agencia, $numero, $saldo)
    {
        $this->titular = $titular;
        $this->agencia = $agencia;
        $this->numero = $numero;
        $this->saldo = $saldo;

        self::$totalDeContas++;
        try {
            if (self::$totalDeContas < 1)
                throw new DivisionByZeroError('Valor inferior a zero!');
            self::$taxaOperação = 1263 / self::$totalDeContas;
        } catch (\DivisionByZeroError $e) {
            echo $e->getMessage();
            exit();
        }
    }

    /**
     * New Constructor of ContaCorrente
     *
     * @param $titular
     * @param $agencia
     * @param $numero
     * @param $saldo
     * @return new Object ContaCorrente
     */
    public static function Conta($titular, $agencia, $numero, $saldo)
    {
        $new = new ContaCorrente($titular, $agencia, $numero, $saldo);
        return $new;
    }

    public function sacar($valor)
    {
        Validacao::verificaNumerico($valor);

        if ($valor > $this->saldo)
            throw new SaldoInsuficienteException('Saldo Insuficiente', $valor, $this->saldo);

        $this->saldo -= $valor;

        return $this;
    }

    public function depositar($valor)
    {
        Validacao::verificaNumerico($valor);

        $this->saldo += $valor;

        return $this;
    }

    public function transferir($valor, ContaCorrente $contaCorrente)
    {
        try {
            Validacao::verificaNumerico($valor);
    
            if ($valor < 0) {
                throw new \Exception('Não é permitido transferências de valor negativo!');
            }
    
            $this->sacar($valor);
            $contaCorrente->depositar($valor);
            return $this;
        } catch (\Exception $e) {
            ContaCorrente::$operacoesNaoRealizadas++;
            throw new OperacaoNaoRealizadaException("Operação não realizada", 55, $e);
        }
    }
    public static function depositarStc(&$conta, $valor)
    {
        $conta->saldo += $valor;
    }

    public static function sacarStc(&$conta, $valor)
    {
        if ($valor > $conta->saldo)
            echo "Impossível sacar o valor de R$ {$valor}. Saldo Insuficiente!" . PHP_EOL;
        else
            $conta->saldo -= $valor;
    }

    public function __toString()
    {
        $conta = ['titular' => $this->titular, 'agencia' => $this->agencia, 'numero' => $this->numero, 'saldo' => number_format($this->saldo, 2, ",", "."), 'totalSaquesNaoPermitidos' => $this->totalSaquesNaoPermitidos];
        return json_encode($conta);
    }
}

File: FormacaoCSharp/FormacaoCSharp/Parte2/P2_06_ByteBank/ContaCorrente.cs
namespace P2_06_ByteBank
{
    class ContaCorrente
    {
        private Cliente _titular;

        public Cliente Titular
        {
            get => _titular;
            set { _titular = value; }
        }

        public int _agencia;
        public int Agencia { get => _agencia; set { _agencia = value; } }
        public int _numero;
        public int Numero { get => _numero; set { _numero = value; } }
        private double saldo = 100;

        public double Saldo
        {
            get => saldo;
            set
            {
                if (value >= 0)
                    saldo = value;
            }
        }

        public bool Sacar(double valor)
        {
            if (saldo < valor)
            {
                return false;
            }

            saldo -= valor;
            return true;
        }
        public void Depositar(double valor) =>
            saldo += valor;

        public bool Transferir(double valor, ContaCorrente contaDestino)
        {
            if (saldo < valor)
                return false;

            saldo -= valor;
            contaDestino.Depositar(valor);

            return true;
        }

        public override string ToString() =>
             string.Format(@"
    {0}
    Agencia: {1},
    Conta: {2},
    Saldo: {3},
", _titular.Nome, _agencia, _numero, saldo);

    }
}